src/methods/join.js

```diff
@@ -806,7 +806,40 @@ export class JoinImplementation {
     c1: CreatedObjects,
     c2: CreatedObjects
   ): PropertyBindings {
+    let joinLeakedObjectProperty = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      if (d1 === undefined) {
+        if (b.object instanceof ObjectValue && c2.has(b.object)) return d2; // no join
+
+        invariant(m2.has(b)); // The alternative path must be the source of leaking
+        return d2;
+      } else if (d2 === undefined) {
+        if (b.object instanceof ObjectValue && c1.has(b.object)) return d1; // no join
+
+        invariant(m1.has(b)); // The consequent path must be the source of leaking
+        return d1;
+      } else {
+        // Make the leak unconditional
+        let isTrue = bv => bv instanceof BooleanValue && bv.equals(realm.intrinsics.true);
```

Comment:
Every value has already has method mightNotBeTrue, also isTrue(bv) => !bv.mightNotBeTrue() but the other way around, so using the latter is probably better.